### PR TITLE
telemetry-gateway: implement Managed Services receiver

### PR DESCRIPTION
**Pull request by bobheadxi**
See https://github.com/sourcegraph/sourcegraph/pull/60991 - this implements support for the new RPC, `RecordEvent`, and the new `ManagedService` identifier type.

To use the new `ManagedService`, appropriately scoped SAMS M2M credentials are required - see go/sams-m2m. This change leverages the SAMS client introduced by the PLG team to do authn/authz based on SAMS M2M auth (`samsm2m. CheckWriteEventsScope`)

To help us track this, I've also added the concept of a "publisher source" mapping to the proto metadata identifier types (e.g. `ManagedService`) - these are used as metric labels.

TODO:

- [x] Add scope to SAMS in line with the new scope strategy: https://github.com/sourcegraph/accounts.sourcegraph.com/pull/624
- [x] Add client credentials to Telemetry Gateway environments: https://github.com/sourcegraph/managed-services/pull/809

## Test plan

- [x] Unit tests
- [x] Create client credentials for Telemetry Gateway Prod and Dev: https://github.com/sourcegraph/accounts.sourcegraph.com/wiki/Operators-Cheat-Sheet#create-a-new-idp-client, configure in `sg.config.overwrite.yaml`
- [x] Create test client credentials for a "telemetry events publisher" (scope `telemetry_gateway::events::write`) - requires https://github.com/sourcegraph/accounts.sourcegraph.com/pull/624
    ```curl -s -X POST \                                                                         
        -H "Authorization: Bearer $MANAGEMENT_SECERET" \
        https://accounts.sourcegraph.com/api/management/v1/identity-provider/clients \
        --data '{"name": "test-telemetry-publisher", "scopes": ["telemetry_gateway::events::write"], "redirect_uris": ["https://sourcegraph.test"]}'```
- [x] Exchange for a token with default scopes using the example CLI: `ACCOUNTS_SERVER=https://accounts.sourcegraph.com go run ./cmd/accounts-client-example create-client-token` (https://github.com/sourcegraph/accounts.sourcegraph.com/pull/643)
- [x] `sg run telemetry-gateway`, to go http://localhost:6080/debug/grpcui/
- [x] Set metadata against Telemetry Gateway service: ![image](https://github.com/sourcegraph/sourcegraph/assets/23356519/dc7bf1b0-78be-4501-b43d-df57f998be9c)
- [x] Make a request as a managed service, a valid event is not needed as we just want to check the auth flow: 
![image](https://github.com/sourcegraph/sourcegraph/assets/23356519/869296f7-f897-4e5b-bbe7-1f7e6e06969d)
- [x] Get scope error, because token does not have prerequisite scope: 
![image](https://github.com/sourcegraph/sourcegraph/assets/23356519/37a49b4e-4e02-450c-8577-09668bd57f00)
- [x] Exchange for a token with the right scopes using the example CLI: `ACCOUNTS_CLIENT_EXAMPLE_CLIENT_SCOPES='telemetry_gateway::events::write' ACCOUNTS_SERVER=https://accounts.sourcegraph.com go run ./cmd/accounts-client-example create-client-token`
- [x] Get a validation error (desired result, because we sent an incomplete event - the key is that the token was accepted)
![image](https://github.com/sourcegraph/sourcegraph/assets/23356519/67ac5c1b-7a6d-4aca-893c-b960b55bbf35)

### CODE REVIEW

**Comment by bobheadxi** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/61022?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#61022** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/61022?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#60991** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/60991?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bobheadxi and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: cmd/telemetry-gateway/internal/server/samsm2m/samsm2m.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		return status.Error(codes.Unauthenticated, "no token header value")
+	}
+
+	// TODO: as part of go/sams-m2m we need to build out a SDK for SAMS M2M
+	// consumers that has a recommended short-caching mechanism. Avoid doing it
+	// for now until we have a concerted effort.
+	result, err := samsClient.IntrospectToken(ctx, token)
+	if err != nil {
+		logger.Error("samsClient.IntrospectToken failed", log.Error(err))
+		return status.Error(codes.PermissionDenied, "unable to validate token")
```
Comment (unknwon) [inline cmd/telemetry-gateway/internal/server/samsm2m/samsm2m.go:47]: Should this code be something like `codes.InternalError`?

File: cmd/telemetry-gateway/internal/server/samsm2m/samsm2m.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/sams"
+)
+
+const requiredSamsScope = "telemetry_gateway::events::write"
+
+// CheckWriteEventsScope ensures the request context has a valid SAMS token with requiredSamsScope.
+// It returns a gRPC status error suitable to be returned directly from an RPC implementation.
+//
+// See: go/sams-m2m
+func CheckWriteEventsScope(ctx context.Context, logger log.Logger, samsClient sams.Client) error {
```
Comment (unknwon) [inline cmd/telemetry-gateway/internal/server/samsm2m/samsm2m.go:24]: Strictly speaking error should be opaque for 4xx/client-side errors, defer to @evict.

**Review by unknwon — APPROVED** [review]

File: cmd/telemetry-gateway/internal/server/samsm2m/samsm2m.go (hunk 1 above)
Comment (bobheadxi) [reply cmd/telemetry-gateway/internal/server/samsm2m/samsm2m.go:47]: Ah, yes - the only permission denied case should be if we retrieve the introspection result and it is an inactive session

File: cmd/telemetry-gateway/internal/server/samsm2m/samsm2m.go (hunk 2 above)
Comment (bobheadxi) [reply cmd/telemetry-gateway/internal/server/samsm2m/samsm2m.go:24]: Opaque as in, I should say if the token is inactive or does not have the prerequisite scope?

Comment (evict) [reply cmd/telemetry-gateway/internal/server/samsm2m/samsm2m.go]: > Opaque as in, I should not say if the token is inactive or does not have the prerequisite scope?
Yup! Looks good now.

**Review by unknwon — APPROVED** [review]
🚀

**Pull request merged**